File: 백준/Bronze/2231. 분해합/분해합.js
const fs = require('fs');
let input = fs.readFileSync("/dev/stdin").toString().trim();

solution(Number(input));

function solution(input) {
  console.log(decomposition(input));
}

function decomposition(input) {
  let len = input.toString().length;

  for (let i = input - len * 9; i < input; i++) {
    let sum = i;
    for (let j = 0; j < i.toString().length; j++) {
      sum += Number(i.toString()[j]);
    }
    if (sum === input) return i;
  }
  return 0;
}

File: 백준/Bronze/2231. 분해합/분해합.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine());
        int N_len = (Integer.toString(N)).length();

        int result = decomposition(N, N_len);

        System.out.print(result);
    }

    public static int decomposition(int N, int length) {
        for (int i = (N - (length*9)); i < N; i++) {
            int sum = 0;
            int num = i;
            while (num!=0) {
                sum += num%10;
                num /= 10;
            }
            if ((sum + i) == N) return i;
        }
        return 0;
    }
}
